From dfcbf0259af14010b778320a8ec87d3c05919c34 Mon Sep 17 00:00:00 2001
From: Oliver Neukum <oneukum@suse.de>
Date: Fri, 2 Aug 2013 14:44:47 +0200
Subject: [PATCH] xhci: set device to D3Cold on shutdown
Patch-Mainline: Never (mainline uses PCI PM)
References: bnc#833097

According to HP it is necessary to put XHCs into
D3 before the system is shut down. If that is not done
the system may spuriously power on again.
Tested by HP.

Signed-off-by: Oliver Neukum <oneukum@suse.de>
---
 drivers/usb/host/xhci.c | 5 ++++-
 1 file changed, 4 insertions(+), 1 deletion(-)

diff --git a/drivers/usb/host/xhci.c b/drivers/usb/host/xhci.c
index f965072..76e9408 100644
--- a/drivers/usb/host/xhci.c
+++ b/drivers/usb/host/xhci.c
@@ -757,9 +757,10 @@ void xhci_stop(struct usb_hcd *hcd)
 void xhci_shutdown(struct usb_hcd *hcd)
 {
 	struct xhci_hcd *xhci = hcd_to_xhci(hcd);
+	struct pci_dev *pdev = to_pci_dev(hcd->self.controller);
 
 	if (xhci->quirks & XHCI_SPURIOUS_REBOOT)
-		usb_disable_xhci_ports(to_pci_dev(hcd->self.controller));
+		usb_disable_xhci_ports(pdev);
 
 	spin_lock_irq(&xhci->lock);
 	xhci_halt(xhci);
@@ -767,6 +768,8 @@ void xhci_shutdown(struct usb_hcd *hcd)
 
 	xhci_cleanup_msix(xhci);
 
+	pci_set_power_state(pdev, PCI_D3cold);
+
 	xhci_dbg(xhci, "xhci_shutdown completed - status = %x\n",
 		    xhci_readl(xhci, &xhci->op_regs->status));
 }
-- 
1.8.3.1

